### ISSUE TITLE
Checking amount of token in liquidity pool

### ISSUE THREAD

**Issue by cmackeen** [shanefontaine/uniswap-python#15]
It is unclear how to use something like get_token_balance() with respect to the Uniswap v2 factory contract. I am trying to view the amount of a token in the Uniswap v2 liquidity pool,  and it seems all I can check is the ERC20 token balance globally.

**Comment by msujithk** [comment]
I added my own function to the library to get amount of tokens in liquidity pool and to get the price of the input token in eth. Here is the code,

def get_token_price_in_eth(self, owner, weth_address,token) -> Wei:
       weth_address = self.w3.toChecksumAddress(weth_address)#WETH
       erc20_weth = self.erc20_contract(weth_address)
       owner = self.w3.toChecksumAddress(owner)#Uni:Token input exchange ex: UniV2:DAI
       weth_balance: int = erc20_weth.functions.balanceOf(owner).call()
       weth_balance = float(self.w3.fromWei(weth_balance,'ether') )
       print (f'WETH quantity in Uniswap Pool = {weth_balance}')
       token_address = self.w3.toChecksumAddress(token) # Like DAI
       erc20 = self.erc20_contract(token_address)
       token_balance: int = erc20.functions.balanceOf(owner).call()
       gwei = 1000000000
       token_balance = float(token_balance / gwei)
        print (f'Token balance in Uniswap Pool = {token_balance}')
        return float(weth_balance/token_balance)  # price of token

**Comment by ErikBjare** [comment]
@cmackeen Yeah, I was lazy and didn't implement `get_ex_token_balance` for Uniswap v2 since I didn't personally have any need for it. 

It should be easy to implement in a similar fashion to what @msujithk suggested (although I haven't tested it), so feel free to submit a PR!

**Comment by yukarinoki** [comment]
@ErikBjare I'd like to work on this issue, is there anyone else working on this issue? Otherwise, would you like to assign me this issue? (Or should I send a pull request without an assign?)

**Comment by ErikBjare** [comment]
@yukarinoki To my knowledge no one is working on it. Go ahead and do your thing, I'll assign you to the issue :)

**Comment by camperrr** [comment]
> I added my own function to the library to get amount of tokens in liquidity pool and to get the price of the input token in eth. Here is the code,
> 
> ```python
> def get_token_price_in_eth(self, owner, weth_address,token) -> Wei:
>     weth_address = self.w3.toChecksumAddress(weth_address)#WETH
>     erc20_weth = self.erc20_contract(weth_address)
>     owner = self.w3.toChecksumAddress(owner)#Uni:Token input exchange ex: UniV2:DAI
>     weth_balance: int = erc20_weth.functions.balanceOf(owner).call()
>     weth_balance = float(self.w3.fromWei(weth_balance,'ether') )
>     print (f'WETH quantity in Uniswap Pool = {weth_balance}')
>     token_address = self.w3.toChecksumAddress(token) # Like DAI
>     erc20 = self.erc20_contract(token_address)
>     token_balance: int = erc20.functions.balanceOf(owner).call()
>     gwei = 1000000000
>     token_balance = float(token_balance / gwei)
>     print (f'Token balance in Uniswap Pool = {token_balance}')
>     return float(weth_balance/token_balance)  # price of token
> ```

How do you find that "owner" knowing only token address?

**Comment by ledzgio** [comment]
Any news about this? I'd like to have a way to get the amount of tokens provided as liquidity in a specific uniswap pool.

**Comment by hosekhoshtaghaza** [comment]
When you create a pair the object returned by the sdk has "tokenAmounts" and "liquidityToken" as two identifiers you can call query. "tokenAmounts" contains the contract addresses for the two tokens you are trying to trade and their relative value and "liquidityToken" contains the contract address for the liquidity pool. To get the balance/rate from the liquidity pool create and instance of the ERC20 contract using the ABI and the address of the liquidity pool (new Contract(erc20ABI, addressOfPool)) and then call the 'balanceOf' function using the token0 and token1 addresses and then process the balances to get the price.

**Comment by 0hax** [comment]
> ```python
>     token_balance: int = erc20.functions.balanceOf(owner).call()
>     gwei = 1000000000
>     token_balance = float(token_balance / gwei)
> ```

Hi,
Can you explain why you are dividing the balance of the token by gwei ? I can't understand this.

**Comment by hosekhoshtaghaza** [comment]
The balance received is wei, replace gwai with how ever many units of wei make up 1 unit of the currency. It just turns the huge numbers in to easier to manage decimals. You should consider using a big number package as there will be implicit rounding issues making your output less precise.

**Comment by Prijwal-Naidu** [comment]
> > I added my own function to the library to get amount of tokens in liquidity pool and to get the price of the input token in eth. Here is the code,
> > ```python
> > def get_token_price_in_eth(self, owner, weth_address,token) -> Wei:
> >     weth_address = self.w3.toChecksumAddress(weth_address)#WETH
> >     erc20_weth = self.erc20_contract(weth_address)
> >     owner = self.w3.toChecksumAddress(owner)#Uni:Token input exchange ex: UniV2:DAI
> >     weth_balance: int = erc20_weth.functions.balanceOf(owner).call()
> >     weth_balance = float(self.w3.fromWei(weth_balance,'ether') )
> >     print (f'WETH quantity in Uniswap Pool = {weth_balance}')
> >     token_address = self.w3.toChecksumAddress(token) # Like DAI
> >     erc20 = self.erc20_contract(token_address)
> >     token_balance: int = erc20.functions.balanceOf(owner).call()
> >     gwei = 1000000000
> >     token_balance = float(token_balance / gwei)
> >     print (f'Token balance in Uniswap Pool = {token_balance}')
> >     return float(weth_balance/token_balance)  # price of token
> > ```
> 
> How do you find that "owner" knowing only token address?

You can generate the token pair using the router address of uniswap (or sushiswap), and the two token addresses.  I answered this question on stackexchange: https://stackoverflow.com/questions/66710238/compute-uniswap-pair-address-via-python/67156507#67156507

**Comment by freaker2k7** [comment]
I also lost you on how to get the owner and why it's the same owner for both tokens/addresses.

**Comment by Prijwal-Naidu** [comment]
> I also lost you on how to get the owner and why it's the same owner for both tokens/addresses.

The 'owner' key is the address for the token pair which is unique for each Exchange.

**Comment by joseddg92** [comment]
So, the function shared on this post can be used to get the liquidity in v2? I am looking to use this in BSC chain.

I already got the fix for constants.py for the BSC chain, and I have also copied the function suggested here, but I do not understand the parameters. Suppose I want to check the liquidity of CAKE in pancakeswap. All I have is the CAKE contract address.

What is owner, weth_addrees and token?

**Comment by joseddg92** [comment]
@msujithk also, I am getting 

AttributeError: 'Uniswap' object has no attribute 'erc20_contract'

Where is that method?

**Comment by joseddg92** [comment]
Sorry for my last two comments, got it figured out.

Working code for BSC:

```
    def get_pool_info(self, owner, weth_address, token) -> Wei:
        weth_address = self.w3.toChecksumAddress(weth_address)  # WETH
        erc20_weth= _load_contract_erc20(self.w3, weth_address)

        owner = self.w3.toChecksumAddress(owner)
        weth_balance: int = erc20_weth.functions.balanceOf(owner).call()
        weth_balance = float(self.w3.fromWei(weth_balance, 'ether'))
        print(f'WETH quantity in Uniswap Pool = {weth_balance}')
        token_address = self.w3.toChecksumAddress(token)  # Like DAI
        erc20 = _load_contract_erc20(self.w3, token_address)

        token_balance: int = erc20.functions.balanceOf(owner).call()
        gwei = 1000000000
        token_balance = float(token_balance / gwei)
        print(f'Token balance in Uniswap Pool = {token_balance}')
        
        return 0.0 if token_balance == 0 else float(weth_balance / token_balance)  # price of token
```

Used like this:

```
# For BSC network
CONTRACTS = {
    "no_liq": "0x0Fc701755e825C488439Df26393505C4AC9e6Fc1",
    "CAKE": "0x0E09FaBB73Bd3Ade0a17ECC321fD13a19e81cE82"
}

PANCAKE_SWAP_FACTORY= "0xcA143Ce32Fe78f1f7019d7d551a6402fC5350c73"
PANCAKE_SWAP_ROUTER = "0x10ED43C718714eb63d5aA57B78B54704E256024E"

for name, addr in CONTRACTS.items():
    pancakeswap.get_pool_info(
        owner=PANCAKE_SWAP_ROUTER,
        weth_address=addr,
        token=addr
    )
```

**Comment by 8ty8t** [comment]
@joseddg92 is there anyway for me to message you about the code above?

**Comment by Mreyna3** [comment]
@joseddg92 Does your code get the amount of lp tokens for a given address

**Comment by hoangduyloc** [comment]
@joseddg92 method is just good for tracking small amounts of tokens, it can not work for most of the tokens. @8ty8t, @Mreyna3. I think we need to figure out some other ways to successfully tracking the liquid pool.

**Comment by jhb10c** [comment]
@Mreyna3, @8ty8t, are you looking to check  LP tokens for V2 or V3 uniswap. For V2, you can check the total LP from the pairs contract function totalSupply(). I think V3 should be roughly similar. If you are still looking for a solution. I can type one up.

**Comment by asdfzxh8** [comment]
joseddg92 no such thing as sorx or fx or etc, cepux, say etc any nmw s perfect

**Comment by 0x6d656d6573** [comment]
> @Mreyna3, @8ty8t, are you looking to check LP tokens for V2 or V3 uniswap. For V2, you can check the total LP from the pairs contract function totalSupply(). I think V3 should be roughly similar. If you are still looking for a solution. I can type one up.

Where can I get the abi for this?
